ngraph/core/reference/src/runtime/reference/random_uniform.cpp

```diff
@@ -0,0 +1,222 @@
+    n_lr.second = prod1.first;
+    counter_lr.first = prod0.second ^ counter_lr.second ^ key_lr.second;
+    counter_lr.second = prod0.first;
+
+    counter = unite_high_low(counter_lr.second, counter_lr.first);
+    n = unite_high_low(n_lr.second, n_lr.first);
+}
+
+void raise_key(uint64_t& key) {
+    auto key_lr = split_high_low(key);
+    key_lr.first += 0x9E3779B9;
+    key_lr.second += 0xBB67AE85;
+    key = unite_high_low(key_lr.second, key_lr.first);
+}
+
+float uint32_to_float(uint32_t x) {
+    uint32_t x_uint32 = (static_cast<uint32_t>(127) << 23) | (x & 0x7fffffu);
+
+    float x_float;
+    memcpy(&x_float, &x_uint32, sizeof(x_uint32));
+    return x_float - 1.0f;
+}
+
+float16 uint32_to_float16(uint32_t x) {
+    uint16_t x_uint16 = static_cast<uint16_t>(x);
+    x_uint16 = (static_cast<uint16_t>(15) << 10) | (x_uint16 & 0x3ffu);
+
+    float16 x_float16;
+    memcpy(&x_float16, &x_uint16, sizeof(x_uint16));
+    return x_float16 - static_cast<float16>(1);
+}
+
+double uint32_to_double(uint32_t x1, uint32_t x2) {
+    uint64_t significant = ((static_cast<uint64_t>(x1) & 0xfffffu) << 32) | static_cast<uint64_t>(x2);
+    uint64_t x_uint64 = ((static_cast<uint64_t>(1023) << 52) | significant);
+
+    double x_double;
+    memcpy(&x_double, &x_uint64, sizeof(x_uint64));
+    return x_double - 1.0;
+}
+
+uint64_t uint32_to_uint64(uint32_t x1, uint32_t x2) {
+    return (static_cast<uint64_t>(x2) << 32) | static_cast<uint64_t>(x1);
+}
+
+bfloat16 uint32_to_bfloat16(uint32_t x) {
+    uint16_t x_uint16 = static_cast<uint16_t>(x);
+    x_uint16 = (static_cast<uint16_t>(127) << 7) | (x_uint16 & 0x7fu);
+
+    bfloat16 x_bfloat16;
+    memcpy(&x_bfloat16, &x_uint16, sizeof(x_uint16));
+    return x_bfloat16 - static_cast<bfloat16>(1);
+}
+
+void run_philox(uint64_t key, uint64_t counter, uint64_t n, size_t n_rounds, std::vector<uint32_t>& res) {
+    for (size_t i = 0; i < n_rounds; i++) {
+        calculate_round(key, counter, n);
+        if (i < n_rounds - 1)
+            raise_key(key);
+    }
+    auto res1 = split_high_low(n);
+    auto res2 = split_high_low(counter);
+    res[0] = res1.first;
+    res[1] = res1.second;
+    res[2] = res2.first;
+    res[3] = res2.second;
+}
+
+void random_uniform(const uint64_t* out_shape,
+                    const char* min_val,
+                    const char* max_val,
+                    char* out,
+                    const Shape& out_shape_shape,
+                    ngraph::element::Type elem_type,
+                    uint64_t seed,
+                    uint64_t seed2) {
+    if (seed == 0 && seed2 == 0) {
+        std::srand(std::time(nullptr));
+        seed = std::rand();
+    }
+    uint64_t key = seed;
+    uint64_t counter = seed2;
+    uint64_t n = 0;
+    size_t shape_count = shape_size(out_shape_shape);
+    size_t elem_count = 1;
+    for (size_t i = 0; i < shape_count; i++) {
+        elem_count *= out_shape[i];
+    }
+    size_t step = elem_type.size() > 4 ? 2 : 4;
```

Comment:
please add comments what here is happening

Reply:
Added comments.